Assets/MRTK/Tools/RuntimeTools/Tools/InputFeatureUsageTool/ListInputFeatureUsages.cs

```diff
@@ -22,104 +23,141 @@ public class ListInputFeatureUsages : MonoBehaviour
         [Tooltip("Used for displaying all detected input source names.")]
         private TextMesh listInputDevicesTextMesh = null;
 
+        [SerializeField]
+        [Tooltip("Used for displaying all detected input source names.")]
+        private GridObjectCollection gridObjectCollection = null;
+
         [SerializeField]
         [Tooltip("Used for displaying data from input.")]
-        private TextMesh[] displayFeatureUsagesTextMeshes = null;
+        private GameObject displayFeatureUsagesPrefab = null;
 
 #if UNITY_2019_3_OR_NEWER
-        private readonly List<InputDevice> controllerInputDevices = new List<InputDevice>();
-        private readonly List<InputDevice> handInputDevices = new List<InputDevice>();
+        private readonly List<InputDevice> leftInputDevices = new List<InputDevice>();
+        private readonly List<InputDevice> rightInputDevices = new List<InputDevice>();
         private readonly List<InputFeatureUsage> featureUsages = new List<InputFeatureUsage>();
+        private readonly List<TextMesh> displayFeatureUsagesTextMeshes = new List<TextMesh>();
+
+        private const float BackingPanelMargin = 0.05f;
+        private const float BackingPanelEntryHeight = 0.03f;
 #endif // UNITY_2019_3_OR_NEWER
 
         private void Update()
         {
-            if (listInputDevicesTextMesh == null || displayFeatureUsagesTextMeshes.Length == 0)
+            if (listInputDevicesTextMesh == null || gridObjectCollection == null || displayFeatureUsagesPrefab == null)
             {
                 return;
             }
 
 #if UNITY_2019_3_OR_NEWER
-            InputDevices.GetDevicesWithCharacteristics(InputDeviceCharacteristics.Controller, controllerInputDevices);
-            InputDevices.GetDevicesWithCharacteristics(InputDeviceCharacteristics.HandTracking, handInputDevices);
+            InputDevices.GetDevicesWithCharacteristics(InputDeviceCharacteristics.Left, leftInputDevices);
+            InputDevices.GetDevicesWithCharacteristics(InputDeviceCharacteristics.Right, rightInputDevices);
 
-            List<InputDevice> inputDevices = controllerInputDevices.Union(handInputDevices).ToList();
+            List<InputDevice> inputDevices = leftInputDevices.Union(rightInputDevices).ToList();
+            int sourceCount = inputDevices.Count;
 
-            listInputDevicesTextMesh.text = $"Detected {inputDevices.Count} input source{(inputDevices.Count > 1 ? "s:" : inputDevices.Count != 0 ? ":" : "s")}\n";
+            listInputDevicesTextMesh.text = $"Detected {sourceCount} input source{(sourceCount > 1 ? "s:" : sourceCount != 0 ? ":" : "s")}\n";
 
-            for (int i = 0; i < displayFeatureUsagesTextMeshes.Length; i++)
+            for (int i = displayFeatureUsagesTextMeshes.Count; i < sourceCount; i++)
+            {
+                displayFeatureUsagesTextMeshes.Add(Instantiate(displayFeatureUsagesPrefab, gameObject.transform).GetComponentInChildren<TextMesh>());
```

Comment:
This seems really odd, we're calling gridObjectCollection.UpdateCollection(); yet we are adding things to displayFeatureUsagesTextMeshes while also iterating through that collection....

Reply:
Small ping on this since I'm not sure I quite understand the architecture here.

Reply:
We iterate from `displayFeatureUsagesTextMeshes.Count` to `sourceCount` to ensure `displayFeatureUsagesTextMeshes` contains enough text meshes to account for all active sources. If they match, no work to be done here! Otherwise, we fill in the list with new text meshes.

`Instantiate(displayFeatureUsagesPrefab, gameObject.transform)` adds any new text meshes under the `gridObjectCollection`'s GameObject, so when we update the collection, it accounts for all the new GOs.

Not sure if that description makes this any clearer haha, but happy to follow up!

Reply:
Didn't realize we were iterating from mesh count to sourceCount! Definitely misread it as a standard iterating through the text meshes.

Approved!